File: iosApp/iosApp/Modules/Degen/Cult/CultProposals/CultProposalsInteractor.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3lib

protocol CultProposalsInteractor: AnyObject {

    func fetchProposals(
        onCompletion: @escaping (Result<[CultProposal], Error>) -> Void
    )
    var hasCultBalance: Bool { get }
}

final class DefaultCultProposalsInteractor {

    private let cultService: CultService
    private let walletSercie: WalletService

    init(
        _ cultService: CultService,
        walletService: WalletService
    ) {
        self.cultService = cultService
        self.walletSercie = walletService
    }
}

extension DefaultCultProposalsInteractor: CultProposalsInteractor {

    func fetchProposals(
        onCompletion: @escaping (Result<[CultProposal], Error>) -> Void
    ) {
        cultService.fetchProposals(handler: onCompletion)
    }
    
    var hasCultBalance: Bool {
        // TODO: Review this when supporting CULT on other networks
        let network = Network.Companion().ethereum()
        let currency = Currency.Companion().cult()
        let balance = walletSercie.balance(network: network, currency: currency)
        guard balance > .zero else { return false }
        return true
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/degenCultProposals/CultProposalsInteractor.kt
package com.sonsofcrypto.web3walletcore.modules.degenCultProposals

import com.sonsofcrypto.web3lib.services.wallet.WalletService
import com.sonsofcrypto.web3lib.types.Currency
import com.sonsofcrypto.web3lib.types.Network
import com.sonsofcrypto.web3lib.utils.BigInt
import com.sonsofcrypto.web3walletcore.services.cult.CultService
import com.sonsofcrypto.web3walletcore.services.cult.CultServiceResponse

interface CultProposalsInteractor {
    @Throws(Throwable::class)
    suspend fun fetch(): CultServiceResponse
    val hasCultBalance: Boolean
}

class DefaultCultProposalsInteractor(
    private val cultService: CultService,
    private val walletService: WalletService,
): CultProposalsInteractor {

    override suspend fun fetch(): CultServiceResponse = cultService.fetch()

    override val hasCultBalance: Boolean get() {
        val balance = walletService.balance(Network.ethereum(), Currency.cult())
        return balance.isGreaterThan(BigInt.zero)
    }
}